todo_app/app.py

```diff
@@ -1,31 +1,120 @@
-from flask import Flask, render_template, request, redirect, url_for
+from functools import wraps
+from os import abort
+
+import requests
+
+from flask import Flask, render_template, request, redirect, url_for, abort
+
+from todo_app import flask_config
+from todo_app.User import User
 from todo_app.mongo_repository import MongoRepository
 from todo_app.viewmodel import ViewModel
+from flask_login import LoginManager, login_required, login_user, current_user
+from oauthlib.oauth2 import WebApplicationClient
 
 
 def create_app():
     app = Flask(__name__)
     app.config.from_object('todo_app.flask_config.Config')
+    config = flask_config.Config()
     mongo_repo = MongoRepository()
 
+    login_manager = LoginManager()
+    app.config['SECRET_KEY'] = config.SECRET_KEY
+    login_manager.init_app(app)
+    login_manager.login_view = "login"
+    client = WebApplicationClient(config.GITHUB_OAUTH_CLIENT_ID)
+
+    @login_manager.unauthorized_handler
+    def unauthenticated():
+        return redirect(client.prepare_request_uri('https://github.com/login/oauth/authorize'))
+
+    # Github OAuth flow when unauthenticated
+    @login_manager.user_loader
+    def load_user(user_id):
+        return mongo_repo.get_user(user_id)
+
+    def authorise_user(role):
+        def authorized(f):
+            @wraps(f)
+            def wrapper(*args, **kwargs):
+                if hasattr(current_user, 'role') and current_user.role != role:
+                    abort(401)
+                return f(*args, **kwargs)
+
+            return wrapper
+
+        return authorized
+
+    @app.route('/login/callback', methods=['GET', 'POST'])
+    def login_callback():
+        token_url, headers, body = client.prepare_token_request('https://github.com/login/oauth/access_token',
+                                                                code=request.args['code'],
+                                                                client_id=config.GITHUB_OAUTH_CLIENT_ID,
+                                                                client_secret=config.GITHUB_OAUTH_CLIENT_SECRET)
+
+        headers = {
+            'Content-Type': 'application/x-www-form-urlencoded',
+            'Accept': 'application/json'
+        }
+
+        response = requests.post(url=token_url, data=body, headers=headers,
+                                 auth=(config.GITHUB_OAUTH_CLIENT_ID, config.GITHUB_OAUTH_CLIENT_SECRET))
+        client.parse_request_body_response(response.content)
+
+        # request for user
+        uri, headers, body = client.add_token('https://api.github.com/user')
+
+        userinfo_response = requests.get(uri, headers=headers)
+
+        user = userinfo_response.json()
+        user_id = user['id']
+
+        # add user to mongo
+        users_count = len(mongo_repo.get_all_users())
+        role = User.ROLE_ADMIN if users_count == 0 else User.ROLE_READER
+        mongo_repo.add_user(user_id, role)
+        login_user(User(user_id, 'reader'))
```

Comment:
```suggestion
        login_user(User(user_id, role))
```
though I don't think this breaks anything